File: services/api/src/endpoints/channels.py
from __future__ import annotations

from typing import Annotated, List

import aio_pika
import pydantic
from fastapi import APIRouter, Depends, Query, WebSocket

from ..core import amqp, rpc
from ..proto import channels_pb2, channels_pb2_grpc
from ..models.adapters import get_converter
from ..models.authorization import AccountToken
from ..models.channels import Channel, Message
from ..models.users import User


__all__ = ("router",)
router = APIRouter(
    prefix="/channels",
)


class __CreateChannelBody(pydantic.BaseModel):
    name: Annotated[str, pydantic.Field(description="The name of the channel")]
    description: Annotated[str, pydantic.Field(description="The description of the channel")]


@router.post(
    "/",
    name="Create a channel",
    description="Create a new channel",
)
async def create_channel(
    body: __CreateChannelBody,
    owner: Annotated[User, Depends(AccountToken.verify)],
) -> Channel:
    stub = channels_pb2_grpc.ChannelServiceStub(await rpc())
    channel: channels_pb2.PChannel = await stub.CreateChannel(
        channels_pb2.PCreateChannelRequest(
            name=body.name,
            description=body.description,
            owner_id=owner.id,
        )
    )

    return Channel(
        id=channel.id,
        name=channel.name,
        description=channel.description,
        owner=User(
            id=channel.owner.id,
            username=channel.owner.username,
            permissions=channel.owner.permissions,
        ),
    )


@router.get(
    "/",
    name="List channels",
    description="Query all channels",
)
async def list_channels() -> List[Channel]:
    stub = channels_pb2_grpc.ChannelServiceStub(await rpc())
    c: channels_pb2.PChannelQueryResult = await stub.Query(None)
    converter = get_converter(channels_pb2.PChannel, Channel)
    return [converter(channel) for channel in c.channels]


@router.websocket(
    "/{channel_id}/ws",
    name="Listen to messages in real-time",
)
async def receive_messages(ws: WebSocket, channel_id: int) -> None:
    await ws.accept()
    channel = await amqp()

    exchange = await channel.declare_exchange(
        "channel-messages",
        aio_pika.ExchangeType.DIRECT,
    )
    queue = await channel.declare_queue()
    await queue.bind(exchange, f"channel-{channel_id}")

    async with queue.iterator() as q:
        data = channels_pb2.PMessage()
        converter = get_converter(channels_pb2.PMessage, Message)

        async for message in q:
            async with message.process():
                data.ParseFromString(message.body)
                await ws.send_json(converter(data).model_dump())


class __CreateMessageBody(pydantic.BaseModel):
    content: Annotated[str, pydantic.Field(description="The content of the message")]


@router.post(
    "/{channel_id}/messages",
    name="Create a message",
    description="Create a new message in a channel",
)
async def create_message(
    channel_id: int,
    body: __CreateMessageBody,
    author: Annotated[User, Depends(AccountToken.verify)],
) -> Message:
    stub = channels_pb2_grpc.ChannelServiceStub(await rpc())
    m: channels_pb2.PMessage = await stub.CreateMessage(
        channels_pb2.PCreateMessageRequest(
            content=body.content,
            author_id=author.id,
            channel_id=channel_id,
        )
    )
    return Message(
        id=m.id,
        content=m.content,
        author=User(
            id=m.author.id,
            username=m.author.username,
            permissions=m.author.permissions,
        ),
        channel=Channel(
            id=m.channel.id,
            name=m.channel.name,
            description=m.channel.description,
            owner=User(
                id=m.channel.owner.id,
                username=m.channel.owner.username,
                permissions=m.channel.owner.permissions,
            ),
        ),
    )


@router.get(
    "/{channel_id}/messages",
    name="List messages",
    description="Query messages in a channel history",
)
async def list_messages(
    channel_id: int,
    *,
    newest: Annotated[bool, Query(description="Sort messages from newest to oldest")] = True,
    before_id: Annotated[int, Query(description="The upper limit of snowflake ID")] = (1 << 53) - 1,  # https://github.com/fastapi/fastapi/discussions/6237
    after_id: Annotated[int, Query(description="The lower limit of snowflake ID")] = 0,
    limit: Annotated[int, Query(description="The maximum number of messages to return (maximum 50)")] = 50,
) -> List[Message]:
    stub = channels_pb2_grpc.ChannelServiceStub(await rpc())
    m: channels_pb2.PHistoryQueryResult = await stub.History(
        channels_pb2.PHistoryQuery(
            id=channel_id,
            newest=newest,
            before_id=before_id,
            after_id=after_id,
            limit=limit,
        )
    )

    converter = get_converter(channels_pb2.PMessage, Message)
    return [converter(message) for message in m.messages]

File: services/frontend/src/api/channels.ts
import JSONBigInt from "json-bigint";

import client from "./client";
import User from "./users";
import Snowflake from "./snowflake";

class FetchMessageQuery {
  public newest: boolean = true;
  public beforeId: bigint = 1n << 63n;
  public afterId: bigint = 0n;
  public limit: number = 50;
}

export class Channel extends Snowflake {
  private static readonly _cache = new Map<bigint, Channel>();
  public static channels: Channel[] = [];

  public readonly name: string;
  public readonly description: string;
  public readonly owner: User;

  public constructor(id: bigint, name: string, description: string, owner: User) {
    super(id);
    this.name = name;
    this.description = description;
    this.owner = owner;
  }

  public async send(content: string): Promise<Message> {
    const response = await client.post<Message>(
      `/channels/${this.id}/messages`,
      { content },
      { transformResponse: [data => data] },
    );
    return response.data;
  }

  public async history(config: FetchMessageQuery): Promise<Message[]> {
    const result: Message[] = [];

    let id = config.newest ? config.beforeId : config.afterId;
    while (config.limit > 0) {
      const limit = Math.min(config.limit, 50);
      config.limit -= limit;

      const response = await client.get<string>(
        `/channels/${this.id}/messages`,
        {
          params: {
            newest: config.newest,
            before_id: config.newest ? id : config.beforeId,
            after_id: config.newest ? config.afterId : id,
            limit,
          },
          transformResponse: [data => data],
        },
      );

      const data = JSONBigInt.parse(response.data) as Message[];
      if (data.length === 0) break;

      for (let i = 0; i < data.length; i++) {
        data[i] = Message.ensure(data[i]);
      }

      if (config.newest) {
        id = data[data.length - 1].id - 1n;
      } else {
        id = data[0].id + 1n;
      }

      result.push(...data);
    }

    return result;
  }

  public poll(handler: (m: Message) => void): WebSocket {
    const ws = client.websocket(`/channels/${this.id}/ws`);
    ws.onmessage = (event) => {
      const message = JSON.parse(event.data) as Message;
      handler(message);
    };
    return ws;
  }

  public static ensure(data: Channel): Channel {
    return new Channel(BigInt(data.id), data.name, data.description, User.ensure(data.owner));
  }

  public static getChannel(id: bigint): Channel | null {
    return Channel._cache.get(id) || null;
  }

  public static async fetchChannel(id: bigint): Promise<Channel> {
    const response = await client.get<string>(
      `/channels/${id}`,
      { transformResponse: [data => data] },
    );
    const channel = JSONBigInt.parse(response.data) as Channel;
    Channel._cache.set(channel.id, channel);
    return channel;
  }

  public static async query(): Promise<Channel[]> {
    const response = await client.get<string>(
      "/channels",
      { transformResponse: [data => data] },
    );
    const channels = JSONBigInt.parse(response.data) as Channel[];
    for (const channel of channels) {
      Channel._cache.set(channel.id, channel);
    }

    channels.sort((a, b) => Number(a.id - b.id));
    Channel.channels = channels;
    return channels;
  }

  public static async create(name: string, description: string): Promise<Channel> {
    const response = await client.post<string>(
      "/channels",
      {
        name,
        description,
      },
      { transformResponse: [data => data] },
    );

    return JSONBigInt.parse(response.data) as Channel;
  }
}

export class Message extends Snowflake {
  public readonly content: string;
  public readonly author: User;
  public readonly channel: Channel;

  public constructor(id: bigint, content: string, author: User, channel: Channel) {
    super(id);
    this.content = content;
    this.author = author;
    this.channel = channel;
  }

  public static ensure(data: Message): Message {
    return new Message(BigInt(data.id), data.content, User.ensure(data.author), Channel.ensure(data.channel));
  }
}
